src/sentry/integrations/slack/event_endpoint.py

```diff
@@ -3,81 +3,114 @@
 import json
 import re
 import six
+from collections import defaultdict
 
 from django.conf import settings
 
 from sentry import http
 from sentry.api.base import Endpoint
+from sentry.incidents.models import Incident
 from sentry.models import Group, Project
 
-from .utils import build_attachment, logger
 from .requests import SlackEventRequest, SlackRequestError
+from .utils import (
+    build_group_attachment,
+    build_incident_attachment,
+    logger,
+)
 
 # XXX(dcramer): this could be more tightly bound to our configured domain,
 # but slack limits what we can unfurl anyways so its probably safe
-_link_regexp = re.compile(r'^https?\://[^/]+/[^/]+/[^/]+/issues/(\d+)')
+_link_regexp = re.compile(r'^https?\://[^/]+/[^/]+/[^/]+/(issues|incidents)/(\d+)')
+
+
+def unfurl_issues(integration, issue_map):
+    results = {
+        g.id: g for g in Group.objects.filter(
+            id__in=set(issue_map.keys()),
+            project__in=Project.objects.filter(
+                organization__in=integration.organizations.all(),
+            )
+        )
+    }
+    if not results:
+        return {}
+
+    return {
+        v: build_group_attachment(results[k]) for k, v in six.iteritems(issue_map)
+        if k in results
+    }
+
+
+def unfurl_incidents(integration, incident_map):
+    results = {
+        i.id: i for i in Incident.objects.filter(
+            id__in=set(incident_map.keys()),
+            organization__in=integration.organizations.all(),
+        )
+    }
+    if not results:
+        return {}
+
+    return {
+        v: build_incident_attachment(results[k]) for k, v in six.iteritems(incident_map)
+        if k in results
+    }
 
 
 # XXX(dcramer): a lot of this is copied from sentry-plugins right now, and will
 # need refactored
 class SlackEventEndpoint(Endpoint):
+    event_handlers = {
+        'issues': unfurl_issues,
+        'incidents': unfurl_incidents,
+    }
+
     authentication_classes = ()
     permission_classes = ()
 
-    def _parse_issue_id_from_url(self, link):
+    def _parse_url(self, link):
         match = _link_regexp.match(link)
         if not match:
-            return
+            return None, None
         try:
-            return int(match.group(1))
+            return match.group(1), int(match.group(2))
         except (TypeError, ValueError):
-            return
+            return None, None
 
     def on_url_verification(self, request, data):
         return self.respond({
             'challenge': data['challenge'],
         })
 
     def on_link_shared(self, request, integration, token, data):
-        issue_map = {}
+        parsed_events = defaultdict(dict)
         for item in data['links']:
-            issue_id = self._parse_issue_id_from_url(item['url'])
-            if not issue_id:
+            event_type, instance_id = self._parse_url(item['url'])
```

Comment:
Do you need to continue supporting unfurls for old links which would be missing the event_type?

Reply:
Event type is just derived from the incoming url, which has always has either `issue` or `incident` in the url

Reply:
But what about old slack messages? Will they continue to work if people scrollback through their slack history?

Reply:
Event type is entirely internal, and we only ever get via parsing a url. It isn't passed to unfurl, it's just used so that we can generate a different format for incident urls.

In terms of historical messages, I'm not sure that slack even hits our endpoint again for them, does it not just cache the unfurl? Regardless, even if it hits our app with the same url again, then we will just parse the url and extract the event_type, so there shouldn't be any issue. We haven't changed the format of the urls, just the regex we use to parse them.